dex/keygen/keygen.go

```diff
@@ -0,0 +1,129 @@
+package keygen
+
+import (
+	"crypto/aes"
+	"crypto/sha256"
+	"fmt"
+
+	"github.com/decred/dcrd/hdkeychain/v3"
+	"github.com/gcash/bchutil/base58"
+)
+
+// RootKeyParams implements hdkeychain.NetworkParams for master
+// hdkeychain.ExtendedKey creation.
+type RootKeyParams struct{}
+
+func (*RootKeyParams) HDPrivKeyVersion() [4]byte {
+	return [4]byte{0x74, 0x61, 0x63, 0x6f} // ASCII "taco"
+}
+func (*RootKeyParams) HDPubKeyVersion() [4]byte {
+	return [4]byte{0x64, 0x65, 0x78, 0x63} // ASCII "dexc"
+}
+
+const extendedKeyLength = 82
+
+// To encrypt with AES, the input length must be a multiple of 16.
+const paddedExtendedKeyLength = 96
+
+// EncryptExtendedKey encrypts an extended key using a password.
+func EncryptExtendedKey(key *hdkeychain.ExtendedKey, pw []byte) ([]byte, error) {
+	pwHash := sha256.Sum256(pw)
+	cipher, err := aes.NewCipher(pwHash[:])
+	if err != nil {
+		return nil, err
+	}
+
+	serializedSeed := base58.Decode(key.String())
+	paddedSerializedSeed := make([]byte, paddedExtendedKeyLength)
+	copy(paddedSerializedSeed[0:extendedKeyLength], serializedSeed)
+	encrypted := make([]byte, paddedExtendedKeyLength)
+	for i := 0; i < paddedExtendedKeyLength; i += 16 {
+		cipher.Encrypt(encrypted[i:i+16], paddedSerializedSeed[i:i+16])
+	}
+
+	return encrypted, nil
+}
+
+// DecryptExtendeKey decrypts an extended key using a password.
+func DecryptExtendedKey(encKey, pw []byte) (*hdkeychain.ExtendedKey, error) {
+	if len(encKey) != paddedExtendedKeyLength {
+		return nil, fmt.Errorf("expected encrypted length to be %v but got %v",
+			paddedExtendedKeyLength, len(encKey))
+	}
+
+	pwHash := sha256.Sum256(pw)
+	cipher, err := aes.NewCipher(pwHash[:])
+	if err != nil {
+		return nil, err
+	}
+
+	decrypted := make([]byte, len(encKey))
+	for i := 0; i < paddedExtendedKeyLength; i += 16 {
+		cipher.Decrypt(decrypted[i:i+16], encKey[i:i+16])
+	}
+	decryptedSeed := base58.Encode(decrypted[0:extendedKeyLength])
+
+	return hdkeychain.NewKeyFromString(decryptedSeed, &RootKeyParams{})
+}
+
+// GenDeepChild derives the leaf of a path of children from a root extended key.
+func GenDeepChild(root *hdkeychain.ExtendedKey, kids []uint32) (*hdkeychain.ExtendedKey, error) {
+	genChild := func(parent *hdkeychain.ExtendedKey, childIdx uint32) (*hdkeychain.ExtendedKey, error) {
+		err := hdkeychain.ErrUnusableSeed
+		for err == hdkeychain.ErrUnusableSeed {
```

Comment:
These are the wrong error codes.  I fixed this recently in https://github.com/decred/dcrdex/pull/1210/files#diff-a0c3b3a03517588f82d5bb3553e3e46ef5377c8bbcb4f8813ebb6d3cb0a6ee86R629-R630